core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -1581,6 +1844,10 @@ class ReplicaManager(val config: KafkaConfig,
       else {
         Set.empty[Partition]
       }
+      stateChangeLogger.info(s"Fencing ${partitionsToBeFenced.size} partition(s)")
+      if (partitionsToBeFenced.nonEmpty) {
+        makeFenced(imageBuilder, partitionsToBeFenced, metadataOffset)
```

Comment:
What happens to the `onLeadershipChange` callback if the partition is fenced? Because we may not be applying the change until later, I don't think we can send the callback through `handleMetadataRecords`. As an alternative, perhaps we need to provide a way to register a (say) `LeadershipListener` which gets notified about all leadership changes.

On a related note, we might need a notification to the `GroupCoordinator` when a partition becomes fenced so that it can stop attempting to write to the log. If we did a `LeadershipListener`, we could have three callbacks: `onBecomeLeader`, `onBecomeFollower`, and `onBecomeFenced`.

Reply:
I deferred the invocations of the `onLeadershipChange` callbacks so they get invoked after we apply the deferred changes.

Here's a counterargument to the suggestion that we notify when partition changes get deferred.  We cannot rely on any received sate that we defer -- it may or may not be the correct state.  We will eventually apply the correct state, but it may or may not be the initial state that we defer for a partition.  I wonder if we want to act upon what could be intermediate state.

What we have may not be the ideal implementation, but I wonder if it at least correct n that the coordinators would not be able to incorrectly append to an offset/txn state partition.

Reply:
I've come around to assuming all bets are off once we are deferring partition state -- its easier to reason about, and the likeliest possibility by far is that we will xoon catch up and enter a consistent state.  This means ideally we should notify right away when a partition first gets deferred state, and the receiver of the callback can decide what to do.  Right now we only have `onLeadershipChange` with an indication if it's a new leader or a new follower.  We might need something new like you suggest.  However, since we are likely to be in a consistent state soon, maybe this is an optimization that can wait -- the "ideal" solution might not be to costly to defer (pun intended).